src/lib.rs

```diff
@@ -105,42 +97,200 @@ impl Default for FxHasher {
 impl FxHasher {
     #[inline]
     fn add_to_hash(&mut self, i: usize) {
-        self.hash = self.hash.rotate_left(5).bitxor(i).wrapping_mul(K);
+        self.hash = self.hash.wrapping_add(i).wrapping_mul(K);
     }
 }
 
 impl Hasher for FxHasher {
     #[inline]
-    fn write(&mut self, mut bytes: &[u8]) {
-        // Ensure all bytes will be consumed
-        const _: () = assert!(size_of::<usize>() <= size_of::<u64>());
-        // Ensure no bytes are discarded by casting to usize
-        const _: () = assert!(size_of::<u32>() <= size_of::<usize>());
-        // Copy the 1 word sized state to a local variable to ensure it
-        // is kept in a register.
-        // See: https://github.com/rust-lang/rustc-hash/pull/34
-        let mut state = self.clone();
-        while let Some(&usize_bytes) = take_first_chunk(&mut bytes) {
-            state.add_to_hash(usize::from_ne_bytes(usize_bytes));
-        }
-        if let Some(&u32_bytes) = take_first_chunk(&mut bytes) {
-            state.add_to_hash(u32::from_ne_bytes(u32_bytes) as usize);
-        }
-        if let Some(&u16_bytes) = take_first_chunk(&mut bytes) {
-            state.add_to_hash(u16::from_ne_bytes(u16_bytes) as usize);
-        }
-        if let Some(&[u8_byte]) = take_first_chunk(&mut bytes) {
-            state.add_to_hash(u8_byte as usize);
-        }
-        *self = state;
+    fn write(&mut self, bytes: &[u8]) {
+        self.write_u64(hash_bytes(bytes));
+    }
+
+    #[inline]
+    fn write_u8(&mut self, i: u8) {
+        self.add_to_hash(i as usize);
+    }
+
+    #[inline]
+    fn write_u16(&mut self, i: u16) {
+        self.add_to_hash(i as usize);
+    }
+
+    #[inline]
+    fn write_u32(&mut self, i: u32) {
+        self.add_to_hash(i as usize);
+    }
+
+    #[inline]
+    fn write_u64(&mut self, i: u64) {
+        self.add_to_hash(i as usize);
+        #[cfg(target_pointer_width = "32")]
+        self.add_to_hash((i >> 32) as usize);
+    }
+
+    #[inline]
+    fn write_u128(&mut self, i: u128) {
+        self.add_to_hash(i as usize);
+        #[cfg(target_pointer_width = "32")]
+        self.add_to_hash((i >> 32) as usize);
+        self.add_to_hash((i >> 64) as usize);
+        #[cfg(target_pointer_width = "32")]
+        self.add_to_hash((i >> 96) as usize);
+    }
+
+    #[inline]
+    fn write_usize(&mut self, i: usize) {
+        self.add_to_hash(i);
+    }
+
+    #[cfg(feature = "nightly")]
+    #[inline]
+    fn write_length_prefix(&mut self, _len: usize) {
+        // Most cases will specialize hash_slice anyway which calls write(),
+        // which encodes the length already.
+    }
+
+    #[cfg(feature = "nightly")]
+    #[inline]
+    fn write_str(&mut self, s: &str) {
+        // We don't need anything special here.
+        self.write(s.as_bytes())
```

Comment:
Similarly: the docs suggest using either write_string_length+write, or write+write_u8(0xff). Why not one of those?

Reply:
Because our `write` implementation already includes the length.

Reply:
Oh, the last line of `hash_bytes`? I see.

The docs for `write` say "You generally should not do length-prefixing as part of implementing this method." Is there a reason why not following this advice gives a better result? If so, another brief comment explaining why would be helpful.

Reply:
> Is there a reason why not following this advice gives a better result?

Because it only takes a single XOR to handle the length in this way, whereas the way the trait suggests costs an extra multiply + add.

Reply:
Ok. Good information for a comment, then.